opm/simulators/wells/WellState.hpp

```diff
@@ -42,45 +44,49 @@ namespace Opm
         typedef std::array< int, 3 >  mapentry_t;
         typedef std::map< std::string, mapentry_t > WellMapType;
 
-        template <class State>
-        void init(const Wells* wells, const State& state)
-        {
-            init(wells, state.pressure());
-        }
-
         /// Allocate and initialize if wells is non-null.
         /// Also tries to give useful initial values to the bhp() and
         /// wellRates() fields, depending on controls.  The
         /// perfRates() field is filled with zero, and perfPress()
         /// with -1e100.
-        void init(const Wells* wells, const std::vector<double>& cellPressures)
+        void init(const std::vector<double>& cellPressures,
+                  const std::vector<Well2>& wells_ecl,
```

Comment:
This (`vector<Well2>`) is a very convenient argument type for what we're doing here, but it does lead to having to call [`Schedule::getWells2(timestep)`](https://github.com/OPM/opm-common/blob/e9367d5116583bce7a57de08d78877bed2940911/opm/parser/eclipse/EclipseState/Schedule/Schedule.hpp#L174) somewhere upstream.  That, in turn, means that we'll be creating a copy of every pertinent `Well2` object each time we call `WellState::init()`.  How often is that?

If that is more than once in a simulation, I think I'd much prefer that this be a vector of well names (and a `Schedule` object) from which to access individual wells as needed.

Reply:
The well code has traditionally been quite difficult work with; this refactor is very welcome - and quite overdue in my opinion. If the `const std::vector<Well2>& wells_ecl` makes this particular code easier to reason about it should in my opinion be merged in the current form.

If it is indeed costly that can be improved in a later PR; that improvement might very well be in the Schedule class.

Reply:
The init() method is called every report step, in fact the whole well machinery is constructed anew every such step. This is not changed here. The vector passed to the init() method is owned by the BlackoilWellModel class, which already in the current master owns such a vector and calls the getWells2() method to build it. It really needs to have it, so there is no extra cost to pass it to WellState::init().

It's worth mentioning a change though: in the current master, the vector of Well2 object held by the BlackoilWellModel would be identical to that returned by getWells2(), i.e. it would contain all wells, including shut wells and those on other processes. That knowledge was instead embedded in the Wells struct. Therefore many of the methods of BlackoilWellModel would iterate over the Wells struct's wells and look up the well by name in the vector of Well2 objects. With this PR, we filter out shut wells and those on other processes upon construction. As a consequence the methods may loop over the vector of Well2 objects directly, simplifying the code.

Reply:
If I understood @atgeirr correctly his plan is that eventually wellState::init() is only called once at the start of the simulation. Currently this init method and the setWellStateFromTarget method in the wellModel does almost the same thing. I think we will benefit from uniting the code. But not in this PR.

Reply:
> [The] knowledge [of shut wells or wells on other processes] was instead embedded in the `Wells` struct. Therefore many of the methods of `BlackoilWellModel` would iterate over the `Wells` struct's wells and look up the well by name in the vector of `Well2` objects. With this PR, we filter out shut wells and those on other processes upon construction. As a consequence the methods may loop over the vector of `Well2` objects directly, simplifying the code.

You can apply the same kind of filtering to a vector of well names.  To be honest I still don't understand what having a `vector<Well2>` allows you to do that a `vector<string>` of well names combined with a `Schedule` object does not.  Unless of course, you actually need to change the well objects themseleves.

Reply:
It is true that one can filter the well names, but that is not the point here (my explanation was just to explain why there was no extra time or memory cost to passing the vector of Well2 objects). In my opinion passing a `Well2` is more compact than passing a well name, a Schedule and a report time step index. It is also more focused: when all we need is the Well2, it is overkill to pass the Schedule.